scripts/install-build-tools.sh

```diff
@@ -14,36 +15,210 @@ if (( $EUID != 0 )); then
+            break
+        fi
+    done
+
+    # install valid python version if not installed yet
+    if [ -z "$PY_INSTALLED" ]; then
+        for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
+            FULL_PY="python${PY_VERS}"
+            # try to install python version from homebrew and verify installation
+            if brew install "${FULL_PY}"; then
+                echo "${FULL_PY} installed successfully"
+                PY_INSTALLED=${PY_VERS}
+                break
+            else
+                # don't exit as we can try on another version if not the last one
+                echo "${FULL_PY} installation failed"
+            fi
+        done
+        echo "Python3 install with homebrew failed, attempted on python: ${PYTHON_VERSIONS[@]}"
+    fi
+
+    # leave if no valid python version is installed
+    if ! which "python${PY_INSTALLED}" &> /dev/null; then
+        echo "Python${PY_INSTALLED} not found in user path"
+        exit 1
+    else
+        # create virtual environment and install python packages for the valid python version
+        PYTHON_PATH=$(which "python${PY_INSTALLED}")
+        create_venv_install_python "${PYTHON_PATH}" ${PY_INSTALLED}
+    fi
 
-    wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | $SUDO apt-key add -
-    $SUDO add-apt-repository "deb http://apt.llvm.org/focal/ llvm-toolchain-focal-14 main"
+# Linux install with apt
+elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    # avoids getting stuck on interactive prompts which is essential for CI/CD
+    export DEBIAN_FRONTEND=noninteractive
+    $SUDO apt update -y
+    $SUDO apt install -y build-essential wget cmake libgtest-dev libbenchmark-dev \
+        lcov git software-properties-common rsync unzip bc
+
+    # Add LLVM GPG key (apt-key is deprecated in Ubuntu 21.04+ so using gpg)
+    wget -qO - https://apt.llvm.org/llvm-snapshot.gpg.key | \
+        gpg --dearmor -o /usr/share/keyrings/llvm-archive-keyring.gpg
+    echo "deb [signed-by=/usr/share/keyrings/llvm-archive-keyring.gpg] http://apt.llvm.org/focal/ llvm-toolchain-focal-14 main" | \
+        $SUDO tee /etc/apt/sources.list.d/llvm.list
+
+    $SUDO apt update -y
     $SUDO apt install -y clang-format-14 clang-tidy-14
-    $SUDO ln -s -f $(which clang-format-14) /usr/local/bin/clang-format
-    $SUDO ln -s -f $(which clang-tidy-14) /usr/local/bin/clang-tidy
+    $SUDO ln -sf $(which clang-format-14) /usr/local/bin/clang-format
+    $SUDO ln -sf $(which clang-tidy-14) /usr/local/bin/clang-tidy
+
+    # check if supported version of python3 is already installed, and save the version
+    for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
+        if which python${PY_VERS} &> /dev/null; then
+            # save the installed version if found
+            PY_INSTALLED=${PY_VERS}
+            break
+        fi
+    done
+
+    # install valid python version if not installed yet
+    if [ -z "$PY_INSTALLED" ]; then
+        for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
```

Comment:
Can we make this either just install one specific version (instead of looping through different versions on failure either exit or retry) or is there a strong argument we can write clearly here in a comment that indicates why we are looping through different versions?

Reply:
Sure, I'll make it try to install just 3.10 if any of the 3.10-3.12 versions not already installed in the user's system. That would be more consistent with the pylint workflow testing 3.10